File: report-generator/src/main/java/com/teamscale/report/ReportUtils.java
package com.teamscale.report;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.teamscale.client.FileSystemUtils;
import com.teamscale.client.TestDetails;
import com.teamscale.report.testwise.ETestArtifactFormat;
import com.teamscale.report.testwise.model.TestExecution;
import com.teamscale.report.testwise.model.TestwiseCoverageReport;
import com.teamscale.report.util.JsonUtils;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Utilities for generating reports. */
public class ReportUtils {

	/** Converts to given test list to a json report and writes it to the given file. */
	public static void writeTestListReport(File reportFile, List<TestDetails> report) throws IOException {
		writeReportToFile(reportFile, report);
	}

	/** Converts to given test execution report to a json report and writes it to the given file. */
	public static void writeTestExecutionReport(File reportFile, List<TestExecution> report) throws IOException {
		writeReportToFile(reportFile, report);
	}

	/** Converts to given testwise coverage report to a json report and writes it to the given file. */
	public static void writeTestwiseCoverageReport(File reportFile, TestwiseCoverageReport report) throws IOException {
		writeReportToFile(reportFile, report);
	}

	/** Converts to given report to a json string. For testing only. */
	public static String getTestwiseCoverageReportAsString(
			TestwiseCoverageReport report) throws JsonProcessingException {
		return JsonUtils.serialize(report);
	}

	/** Writes the report object to the given file as json. */
	private static <T> void writeReportToFile(File reportFile, T report) throws IOException {
		File directory = reportFile.getParentFile();
		if (!directory.isDirectory() && !directory.mkdirs()) {
			throw new IOException("Failed to create directory " + directory.getAbsolutePath());
		}
		JsonUtils.serializeToFile(reportFile, report);
	}

	/** Recursively lists all files in the given directory that match the specified extension. */
	public static <T> List<T> readObjects(ETestArtifactFormat format, Class<T[]> clazz,
										  List<File> directoriesOrFiles) throws IOException {
		List<File> files = listFiles(format, directoriesOrFiles);
		ArrayList<T> result = new ArrayList<>();
		for (File file : files) {
			T[] t = JsonUtils.deserializeFile(file, clazz);
			if (t != null) {
				result.addAll(Arrays.asList(t));
			}
		}
		return result;
	}

	/** Recursively lists all files of the given artifact type. */
	public static List<File> listFiles(ETestArtifactFormat format, List<File> directoriesOrFiles) {
		List<File> filesWithSpecifiedArtifactType = new ArrayList<>();
		for (File directoryOrFile : directoriesOrFiles) {
			if (directoryOrFile.isDirectory()) {
				filesWithSpecifiedArtifactType.addAll(FileSystemUtils
						.listFilesRecursively(directoryOrFile, file -> fileIsOfArtifactFormat(file, format)));
			} else if (fileIsOfArtifactFormat(directoryOrFile, format)) {
				filesWithSpecifiedArtifactType.add(directoryOrFile);
			}
		}
		return filesWithSpecifiedArtifactType;
	}

	private static boolean fileIsOfArtifactFormat(File file, ETestArtifactFormat format) {
		return file.isFile() &&
				file.getName().startsWith(format.filePrefix) &&
				FileSystemUtils.getFileExtension(file).equalsIgnoreCase(format.extension);
	}
}

File: report-generator/src/main/kotlin/com/teamscale/report/ReportUtils.kt
package com.teamscale.report

import com.fasterxml.jackson.core.JsonProcessingException
import com.teamscale.client.FileSystemUtils
import com.teamscale.client.JsonUtils
import com.teamscale.client.TestDetails
import com.teamscale.report.testwise.ETestArtifactFormat
import com.teamscale.report.testwise.model.TestExecution
import com.teamscale.report.testwise.model.TestwiseCoverageReport
import java.io.File
import java.io.IOException
import java.util.*

/** Utilities for generating reports.  */
object ReportUtils {
	/** Converts to given test list to a json report and writes it to the given file.  */
	@Throws(IOException::class)
	@JvmStatic
	fun writeTestListReport(reportFile: File, report: List<TestDetails>) {
		writeReportToFile(reportFile, report)
	}

	/** Converts to given test execution report to a json report and writes it to the given file.  */
	@Throws(IOException::class)
	@JvmStatic
	fun writeTestExecutionReport(reportFile: File, report: List<TestExecution>) {
		writeReportToFile(reportFile, report)
	}

	/** Converts to given testwise coverage report to a json report and writes it to the given file.  */
	@Throws(IOException::class)
	fun writeTestwiseCoverageReport(reportFile: File, report: TestwiseCoverageReport) {
		writeReportToFile(reportFile, report)
	}

	/** Converts to given report to a json string. For testing only.  */
	@JvmStatic
	@Throws(JsonProcessingException::class)
	fun getTestwiseCoverageReportAsString(
		report: TestwiseCoverageReport
	) = JsonUtils.serialize(report)

	/** Writes the report object to the given file as json.  */
	@Throws(IOException::class)
	private fun <T> writeReportToFile(reportFile: File, report: T) {
		val directory = reportFile.getParentFile()
		if (!directory.isDirectory() && !directory.mkdirs()) {
			throw IOException("Failed to create directory " + directory.absolutePath)
		}
		JsonUtils.serializeToFile(reportFile, report)
	}

	/** Recursively lists all files in the given directory that match the specified extension.  */
	@Throws(IOException::class)
	@JvmStatic
	fun <T> readObjects(
		format: ETestArtifactFormat,
		clazz: Class<Array<T>>,
		directoriesOrFiles: List<File>
	) = listFiles(format, directoriesOrFiles)
		.map { JsonUtils.deserializeFile(it, clazz) }
		.flatMap { listOf(*it) }

	/** Recursively lists all files of the given artifact type.  */
	@JvmStatic
	fun listFiles(
		format: ETestArtifactFormat,
		directoriesOrFiles: List<File>
	) = directoriesOrFiles.flatMap { directoryOrFile ->
		when {
			directoryOrFile.isDirectory() -> {
				directoryOrFile.walkTopDown().filter { it.isOfArtifactFormat(format) }.toList()
			}

			directoryOrFile.isOfArtifactFormat(format) -> {
				listOf(directoryOrFile)
			}

			else -> emptyList()
		}
	}

	private fun File.isOfArtifactFormat(format: ETestArtifactFormat) =
		isFile() &&
				getName().startsWith(format.filePrefix) &&
				FileSystemUtils.getFileExtension(this).equals(format.extension, ignoreCase = true)
}
